exporters/otlp/otlpmetric/otlpmetricgrpc/client.go

```diff
@@ -0,0 +1,212 @@
+	metricpb "go.opentelemetry.io/proto/otlp/metrics/v1"
+)
+
+// New returns an OpenTelemetry metric Exporter. The Exporter can be used with
+// a PeriodicReader to export OpenTelemetry metric data to an OTLP receiving
+// endpoint using gRPC.
+//
+// If an already established gRPC ClientConn is not passed in options using
+// WithGRPCConn, a connection to the OTLP endpoint will be established based
+// on options. If a connection cannot be establishes in the lifetime of ctx,
+// an error will be returned.
+func New(ctx context.Context, options ...Option) (metric.Exporter, error) {
+	c, err := newClient(ctx, options...)
+	if err != nil {
+		return nil, err
+	}
+	return otlpmetric.New(c), nil
+}
+
+type client struct {
+	metadata      metadata.MD
+	exportTimeout time.Duration
+	requestFunc   retry.RequestFunc
+
+	// ourConn keeps track of where conn was created: true if created here in
+	// NewClient, or false if passed with an option. This is important on
+	// Shutdown as the conn should only be closed if we created it. Otherwise,
+	// it is up to the processes that passed the conn to close it.
+	ourConn bool
+	conn    *grpc.ClientConn
+	msc     colmetricpb.MetricsServiceClient
+}
+
+// newClient creates a new gRPC metric client.
+func newClient(ctx context.Context, options ...Option) (otlpmetric.Client, error) {
+	cfg := oconf.NewGRPCConfig(asGRPCOptions(options)...)
+
+	c := &client{
+		exportTimeout: cfg.Metrics.Timeout,
+		requestFunc:   cfg.RetryConfig.RequestFunc(retryable),
+		conn:          cfg.GRPCConn,
+	}
+
+	if len(cfg.Metrics.Headers) > 0 {
+		c.metadata = metadata.New(cfg.Metrics.Headers)
+	}
+
+	if c.conn == nil {
+		// If the caller did not provide a ClientConn when the client was
+		// created, create one using the configuration they did provide.
+		conn, err := grpc.DialContext(ctx, cfg.Metrics.Endpoint, cfg.DialOptions...)
+		if err != nil {
+			return nil, err
+		}
+		// Keep track that we own the lifecycle of this conn and need to close
+		// it on Shutdown.
+		c.ourConn = true
+		c.conn = conn
+	}
+
+	c.msc = colmetricpb.NewMetricsServiceClient(c.conn)
+
+	return c, nil
+}
+
+// ForceFlush does nothing, the client holds no state.
+func (c *client) ForceFlush(ctx context.Context) error { return ctx.Err() }
+
+// Shutdown shuts down the client, freeing all resource.
+//
+// Any active connections to a remote endpoint are closed if they were created
+// by the client. Any gRPC connection passed during creation using
+// WithGRPCConn will not be closed. It is the caller's responsibility to
+// handle cleanup of that resource.
+func (c *client) Shutdown(ctx context.Context) error {
+	// The otlpmetric.Exporter synchronizes access to client methods and
+	// ensures this is called only once. The only thing that needs to be done
+	// here is to release any computational resources the client holds.
+
+	c.metadata = nil
+	c.requestFunc = nil
+	c.msc = nil
+
+	err := ctx.Err()
```

Comment:
gRPC `Close()` is probably always instantaneous. It'd be better to just always call it to clean up the connection (if `ourConn` is true).

Reply:
> gRPC `Close()` is probably always instantaneous.

Is it? It acquires a lock, and waits for the tear-down of all underlying connections:

https://github.com/grpc/grpc-go/blob/d5dee5fdbdeb52f6ea10b37b2cc7ce37814642d7/clientconn.go#L1014-L1062

If the user has included a deadline to this call, making the additional call to Close beyond the deadline would be counter to how the context is defined:

> work done on behalf of this context should be canceled

Reply:
Well, leaking the underlying gRPC connection with all the resources is not better than blocking. 🤷 It does lock a mutex, what I wanted to say is that it probably don't block for a significant duration i.e. presumably it does an abrupt connection close for all underlying streams and TCP connections and doesn't wait for the application code using them to handle that. I'm sorry, I don't have time now to dig into this properly, but not closing the gRPC conn just doesn't feel like the right thing to do. E.g. it's a typical pattern to `defer` the close then use a context. You don't care if the context was cancelled, you want the connection to be torn down anyway.

Reply:
> work done on behalf of this context should be canceled

I'm pretty sure that don't imply application should leak resources when a context is cancelled :)